spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/CassandraJoinRDD.scala

```diff
@@ -0,0 +1,152 @@
+import com.datastax.spark.connector.util.CountingIterator
+import com.datastax.spark.connector.writer._
+import org.apache.spark.rdd.RDD
+import org.apache.spark.{Partition, TaskContext}
+
+import scala.reflect.ClassTag
+
+// O[ld] Is the type of the left Side RDD, N[ew] the type of the right hand side Results
+/**
+ * An RDD that will do a selecting join between `prev:RDD` and the specified Cassandra Table
+ * This will perform individual selects to retrieve the rows from Cassandra and will take
+ * advantage of RDD's that have been partitioned with the [[com.datastax.spark.connector.rdd.partitioner.ReplicaPartitioner]]
+ */
+class CassandraJoinRDD[O, N] private[connector](prev: RDD[O],
+                                                val keyspaceName: String,
+                                                val tableName: String,
+                                                val connector: CassandraConnector,
+                                                val columnNames: ColumnSelector = AllColumns,
+                                                val joinColumns: ColumnSelector = PartitionKeyColumns,
+                                                val where: CqlWhereClause = CqlWhereClause.empty,
+                                                val readConf: ReadConf = ReadConf())
+                                               (implicit oldTag: ClassTag[O], val rct: ClassTag[N],
+                                                @transient val rwf: RowWriterFactory[O], @transient val rtf: RowReaderFactory[N])
+  extends CassandraRDD[(O, N)](prev.sparkContext, prev.dependencies) with CassandraReader[N] {
+
+  //Make sure copy operations make new CJRDDs and not CRDDs
+  override protected def copy(columnNames: ColumnSelector = columnNames,
+                              where: CqlWhereClause = where,
+                              readConf: ReadConf = readConf, connector: CassandraConnector = connector) =
+    new CassandraJoinRDD[O, N](prev, keyspaceName, tableName, connector, columnNames, joinColumns, where, readConf).asInstanceOf[this.type]
+
+  lazy val joinColumnNames: Seq[NamedColumnRef] = joinColumns match {
+    case AllColumns => throw new IllegalArgumentException("Unable to join against all columns in a Cassandra Table. Only primary key columns allowed")
+    case PartitionKeyColumns => tableDef.partitionKey.map(col => col.columnName: NamedColumnRef).toSeq
+    case SomeColumns(cs@_*) => {
+      checkColumnsExistence(cs)
+      checkValidJoin(cs)
+    }
+  }
+
+  protected def checkValidJoin(columns: Seq[NamedColumnRef]): Seq[NamedColumnRef] = {
+    val allColumnNames = tableDef.allColumns.map(_.columnName).toSet
+    val regularColumnNames = tableDef.regularColumns.map(_.columnName).toSet
+    val partitionKeyColumnNames = tableDef.partitionKey.map(_.columnName).toSet
+    val joinColumnNames = columns.map(_.columnName).toSet
+
+    def checkSingleColumn(column: NamedColumnRef) = {
```

Comment:
The primary purpose of this method is to eventually throw an exception. Therefore why not make it return Unit?